### PR TITLE
Enigma post

### PR DESCRIPTION
**Pull request by emedvedev**
The "Enigma emulator in Go" blog post. Sorry I am a little late with it!

It is quite lengthy, but mostly because rotor cipher machines are far from being the easiest subject in the world. :) 

Feel free to cut and edit anything you want; I'll go over it a couple more times myself before publication. Also, English is not my native language, so I would really ask someone to review the post for spelling errors (sorry to be a bother!).

Images are uploaded, everything is formatted according to the requirements (well, except I used 80 characters for the Markdown width, because 72 is a little bit inhumane in 2016). Hugo build is successful, and I checked that the post looks fine on localhost.

Thanks for inviting me to write. :)

### CODE REVIEW

File: content/advent-2016/enigma-emulator-in-go.md [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+
+Footnote: using runes would be more semantically correct than bytes, I guess, but we only encrypt ASCII, so it's fine. The real reason behind using bytes is that they've turned out to work a little faster when I had to run eleven million Enigma configurations.
+
+
+
+# Initial configuration
+
+![](/postimages/advent-2016/enigma-emulator-in-go/keysheet.jpg)
+
+To decrypt an Enigma-encoded message, you must know the configuration that was used to encrypt it. Enigma configurations were distributed on key sheets, like the one above, and changed daily. For obvious reasons, key sheets were considered top secret information and highly protected. They were only given to officers, and often required to set personally and lock the insides of the machine, so that even the operators did not know the settings.
```
Comment (bradleyfalzon) [inline content/advent-2016/enigma-emulator-in-go.md]: Should `to set personally` be `to be set personally`?

File: content/advent-2016/enigma-emulator-in-go.md [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+	ID          string
+	StraightSeq [26]int
+	ReverseSeq  [26]int
+	Turnover    []int
+
+	Offset int
+	Ring   int
+}
+
+type Rotors []*Rotor
```
Comment (bradleyfalzon) [inline content/advent-2016/enigma-emulator-in-go.md]: nit: are you certain `[]*Rotor` is the best, I'm curious as you previously mentioned a performance test, was this the result of that test?

File: content/advent-2016/enigma-emulator-in-go.md [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+	NewRotor("JPGVOUMFYQBENHZRDKASXLICTW", "VI", "ZM"),
+	NewRotor("NZJHGRCXMYSWBOUFAIVLPEKQDT", "VII", "ZM"),
+	NewRotor("FKQHTLXOCBJSPDZRAMEWNIUYGV", "VIII", "ZM")
+}
+```
+
+Finally, the method performing the substitution. We need to account for the current position and the ring settings — this is one of the examples where thinking of letters as their indexes helps us avoid unnecessary conversions:
+
+```
+func (r *Rotor) Step(letter *int, invert bool) {
```
Comment (bradleyfalzon) [inline content/advent-2016/enigma-emulator-in-go.md]: I'd think returning the stepped letter is better than changing the letter in place, but if this is better for the application, then I understand.

File: content/advent-2016/enigma-emulator-in-go.md [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+You could also use a real Enigma machine, of course: after all, there is nothing better than original, and it will only run you down [a meager sum of $365,000](https://www.theguardian.com/world/2015/oct/23/rare-nazi-enigma-machine-sold-at-auction-for-world-record-365000). You can't afford _not_ to buy it!
+
+And of course, I just cannot recommend [the Pringles Enigma](https://fhcouk.files.wordpress.com/2012/05/pringlesenigma3a4.pdf) enough as a cheaper and more functional alternative. You can eat and encode with the same device! This is the future, right here.
+
+And finally, a simple exercise for the reader:
+
+```
+III II IV | 5 10 18 | A A A | AE DQ RC VB MT OG PF YL JW IZ
+
+PJHLFULLUECCPFLCIVPMFDAWJCWANLVXAIXFHMACNLVNCSXOIXFUTGWXSRULRTXPOIPUINCYOGWKGZAZDMVPOUIDCRSCHSZCNTFJADAVIKOGSYAJGAFNELPOMBMTXEXVAREVMSBNHLJFEGZ
```
Comment (bradleyfalzon) [inline content/advent-2016/enigma-emulator-in-go.md]: SZIIIQJJCTNWBMWWIDNGRROKKVLXSCWOVPMLWFZMQQTLMNXSBQITTEGPAPADGPXZYATFSEMDSDXQXBTAAAPZDIZENBJ

File: content/advent-2016/enigma-emulator-in-go.md (hunk 2 above)
Comment (emedvedev) [inline content/advent-2016/enigma-emulator-in-go.md]: Not certain, so let me try to do it a little differently and check :) I do agree that `[]Rotor` would seem more logical, and I don't entirely remember the full train of thought (and benchmarks) that turned it into the list of pointers.

File: content/advent-2016/enigma-emulator-in-go.md (hunk 3 above)
Comment (emedvedev) [inline content/advent-2016/enigma-emulator-in-go.md]: It was another pprof-induced change, but I'll try to make it return the letter again and see if it doesn't slow down things too much. It'd certainly be less confusing that way.

**Comment by emedvedev** [comment]
@bradleyfalzon fixed most of your comments (and found a few more typos in the process), except the two about the code itself. I'll check those in a bit, too.

WXJJBDRJWTEYACEGIGMJGBPWJVIASCWKDL

**Comment by emedvedev** [comment]
@bradleyfalzon addressed both code comments, too. Feel free to review again :)

File: content/advent-2016/enigma-emulator-in-go.md [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+This story begins the day after I got home after giving a talk at the amazing
+[DevFest Siberia](http://devfest.gdg.org.ru/). Shortly after my weekly fix of
+Westworld, a strange nagging feeling appeared — like the one you get from
+unpaid bills, a postcard that you forgot to send, or a particularly nasty API
+endpoint that you were supposed to refactor a year ago but then, well, I mean,
+it works, right?
+
+Then I remembered. I emptied out my suitcase and found the cause of the
+uneasiness: two stickers with goofy cartoon gophers on them. The first one was
+really enjoying his beer. The other one has just put on a wizard hat — possibly
```
Comment (Pisush) [inline content/advent-2016/enigma-emulator-in-go.md]: maybe `their` beer? I think the buddy is genderless

Comment (emedvedev) [inline content/advent-2016/enigma-emulator-in-go.md]: Good catch! Fixed, thanks :)

**Pull request merged**